Niamh25/DOC-5563/Docs-crypto-api (#1592)
corda/corda-docs-portal: content/en/platform/corda/5.0/developing-applications/api/api-crypto.md

Author: Yes, while speaking to Kyriakos he suggested that this page should align more with the rest of the API pages and to look more like: https://nadine.preview.docs.r3.com/en/platform/corda/5.0/developing-applications/api/application.html. Do you think the code snippets should be re-added?

Reviewer: I'm not sure ...
@nadinequinr3 / @paulmoloneyr3 / @colin-hillis  - please chime in ...
I definitely think we should be consistent across the docs portal ...

Reviewer 2: I haven't been following the history of this PR as it was marked as `don't review`. I think if we have code snippets that would be useful to developers we should publish them (with explanations). It could well be the case that some of the other API descriptions are missing code snippets.
If the issue is that it will take a while to prepare the code snippets and surrounding info, it might make sense to go basic with this PR to get something up on the site for the API and then add further content in a future PR?

    parent: corda5-api
    weight: 5000
section_menu: corda5
---
# net.corda.v5.crypto

> The `crypto` package provides services and types for performing cryptographic operations. The following services are available:

Reviewer 3: Only just noticed this, the `crypto` package only contains types and not services.

Perhaps we could say "types used by services in `net.corda.v5.application.crypto`" as a first link between the two.

    parent: corda5-api
    weight: 5000
section_menu: corda5
---
# net.corda.v5.crypto

> The `crypto` package provides services and types for performing cryptographic operations. The list of services available are:

Reviewer 2: ```suggestion
The `crypto` package provides services and types for performing cryptographic operations. The following services are available:
```

> * `CompositeKey` is a tree data structure that enables the representation of composite public keys, which are used to represent the signing requirements for multi-signature scenarios.

Reviewer 3: I understand this is as mentioned on the kDoc but sounds rather an implementation detail. Maybe let's leave as is for now (to be aligned with the kDoc as well) but maybe we should update the kDoc and this to begin with something more generic/ simple of what a `CompositeKey` is.

> * The `CompositeKey`is a tree data structure that enables the representation of composite public keys, which are used to represent the signing requirements for multi-signature scenarios.

Reviewer 4: ```suggestion
* The `CompositeKey` is a tree data structure that enables the representation of composite public keys, which are used to represent the signing requirements for multi-signature scenarios.
```

> * The `CompositeKey` is a tree data structure that enables the representation of composite public keys, which are used to represent the signing requirements for multi-signature scenarios.

Reviewer 2: ```suggestion
* `CompositeKey` is a tree data structure that enables the representation of composite public keys, which are used to represent the signing requirements for multi-signature scenarios.
```

> * `DigestAlgorithmName` is used in the Corda hashing API.

Reviewer 3: I think this API is quite trivial, so maybe let's remove.

> * The `DigestAlgorithmName` is to be used in Corda hashing API.

Reviewer 2: ```suggestion
* `DigestAlgorithmName` is used in the Corda hashing API.
```

> * The `DigitalSignature` is for identifying who is the owner of the signing key used to create this signature.

Reviewer 2: ```suggestion
* `DigitalSignature` is used to identify the owner of the signing key used to create the signature.
```

> * The `SecureHash` is a cryptographically secure hash value, computed by a specified digest algorithm.

Reviewer 2: ```suggestion
* `SecureHash` is a cryptographically secure hash value, computed by a specified digest algorithm.
```

> * The `SignatureSpec`  is a digital signature scheme.

Reviewer 4: ```suggestion
* The `SignatureSpec` is a digital signature scheme.
```

> * The `SignatureSpec` is a digital signature scheme.

Reviewer 2: ```suggestion
* `SignatureSpec` is a digital signature scheme.
```

Reviewer 2: Maybe add a cross-ref to below section?

> ## Implementing Signature Schemes

Reviewer 3: This should rather be "Implemented" or maybe "Offered" I think.

Also maybe not for this PR but I am not sure this should go here? This sounds to me more like implementation and this page is about API. Maybe we could consider moving it elsewhere? The only thing we need at this point is mention `SignatureSpec` which we do further above.

> Corda supports the following `SignatureSpecs` (signature schemes) for creating the following objects:

Reviewer 3: ```suggestion
Corda supports the following `SignatureSpec`s (signature schemes) for creating the following objects:
```

Reviewer 3: Whatever is in ` needs to be matching code.

     * as a single element iterable. 
     * <p>
     * Since we do not define composite keys as acceptable on the second argument of this function, this relation
     * is not reflexive, not symmetric and not transitive. 
     *
     * @param key The key to be checked whether it is being fulfilled by {@code otherKey}.
     * @param otherKey The key against which the {@code key} is being checked for fulfilment.
     */
    public static boolean isKeyFulfilledBy(@NotNull PublicKey key, @NotNull PublicKey otherKey) {
        return isKeyFulfilledBy(key,
                Collections.singleton(otherKey));
    }
}

```

## MessageAuthenticationCode.java

```kotlin

package net.corda.v5.crypto;

import org.jetbrains.annotations.NotNull;

public final class MessageAuthenticationCode {
    private MessageAuthenticationCode() {}

    /**
     * Constant specifying the HMAC SHA-256 algorithm.
     */
    @NotNull
    public static final String HMAC_SHA256_ALGORITHM = "HmacSHA256";

    /**
     * Constant specifying the HMAC SHA-512 algorithm.
     */
    @NotNull
    public final static String HMAC_SHA512_ALGORITHM = "HmacSHA512";
}

```

## SecureHash.java

> ```kotlin

Reviewer: should this be ```java ?
Please review everywhere ...

* GOST3411withGOST3410

Use `SignatureSpecService` to retrieve the `SignatureSpec`.

> Initially the above list of signature spec was available to users. However, they could be passing the wrong signature spec for a signing key type (e.g. an RSA key with an SHA256withECDSA), which would lead to an error when attempting to generate the signature.

Reviewer 4: ```suggestion
Initially the above list of signature spec was available to users. However, they could be passing the wrong signature spec for a signing key type (for example, an RSA key with an SHA256withECDSA), which would lead to an error when attempting to generate the signature.
```

> 

Reviewer 2: ```suggestion
```

> Instead SignatureSpecService was introduced which takes in a key (and a digest algorithm optionally) and returns a default signature spec for those.

Reviewer 4: ```suggestion
Instead, `SignatureSpecService` was introduced. It takes in a key (and a digest algorithm optionally) and returns a default signature spec for those.
```
Also, not sure what "those" refers to. Users?

> Instead, `SignatureSpecService` was introduced. It takes in a key (and a digest algorithm optionally) and returns a default signature spec for those.

Reviewer 2: These two lines read like a Release Note and I don't think belong here.

`SignatureSpecService` ensures that you do not pass the wrong signature spec for a signing-key type. It takes in a key and, optionally, a digest algorithm, and returns the appropriate default signature spec.

        return otherKey.weight == weight;
    }
}

```

## CordaOID.java

{{< tabs name="CordaOID">}}

> {{% tab name="Kotlin" %}}

Reviewer: where is the removal of the `{{< tabs ...` and `{{% tab ...` enclosures ?

Author: There was no enclosures, as this content is still in draft mode and I originally thought there would be Kotlin and java  I had tabs in as a place stopper but since it's just java right now I removed it!
